ts/main.ts

```diff
@@ -69,3 +87,75 @@ function resetBgs(): void {
   $titleInput.classList.remove('value-changed');
   $imgInput.classList.remove('value-changed');
 }
+
+function renderEntry(entry: Entry): HTMLElement {
+  const $li = document.createElement('li');
+  const $row = document.createElement('div');
+  $row.className = 'row';
+  $li.appendChild($row);
+  const $entryImg = document.createElement('img');
+  $entryImg.className = 'entry-img';
+  const $entryH2 = document.createElement('h2');
+  const $entryP = document.createElement('p');
+  const $columnHalf1 = document.createElement('div');
+  $columnHalf1.className = 'column-half';
+  $row.appendChild($columnHalf1);
+  const $columnHalf2 = document.createElement('div');
+  $columnHalf2.className = 'column-half';
+  $row.appendChild($columnHalf2);
+  $entryImg.setAttribute('src', entry.imgUrl);
+  $columnHalf1.appendChild($entryImg);
+  $entryH2.textContent = entry.title;
+  $columnHalf2.appendChild($entryH2);
+  $entryP.textContent = entry.notes;
+  $columnHalf2.appendChild($entryP);
+  return $li;
+}
+
+document.addEventListener('DOMContentLoaded', generatePastEntries);
+
+function generatePastEntries(): void {
+  if (!$ulForEntries) throw new Error('$ulForEntries not found!');
+  for (const entry of data.entries) {
+    const $liToAppend = renderEntry(entry);
+    $ulForEntries.appendChild($liToAppend);
+  }
+  if (!$noEntries) throw new Error('$noEntries not found!');
+  if (
+    $ulForEntries.childNodes.length > 1 &&
+    !$noEntries.classList.contains('hidden')
+  )
+    toggleNoEntries();
+  viewSwap(data.view);
+}
+
+function toggleNoEntries(): void {
+  if (!$noEntries) throw new Error('$noEntries not found!');
+  $noEntries.classList.toggle('hidden');
+}
+
+function viewSwap(newView: string): void {
+  data.view = newView;
+  for (const container of allViews) {
+    if (!container) throw new Error('container not found!');
```

Comment:
Minor thing but you would want to be a little more specific here even if it says something like `'entry or form view container not found'`. By just saying `container`, if I saw that error message I would not know which container since we have a class named `container` as well.